File: legacy/src/game/pinball-knight/maze/floor-metrics.ts
/**
 * FLOOR METRICS — how we judge a generated floor, as numbers.
 *
 * Every generator change before this one was argued from the source: read the
 * pass, reason about what it must produce, ship. That is how a cap that binds
 * from level 1 survives being called "scales with floor area" for two waves,
 * and how five floor archetypes can shape a grid the live path throws away
 * while the descent card announces them by name.
 *
 * So: measure the output instead. This module is the metric half of
 * `.agents/game-dev-rules/procedural-level-generation.md` §6 — a pure function
 * from a finished grid to the numbers that decide whether it is a good floor —
 * and `checkFloor` is the constraint half (§4), which the smoke gate asserts on
 * and any tuning script can call.
 *
 * Deliberately DOM-, three- and content-free: it reads only the tile grid, the
 * endpoints and (optionally) the track mask, so it can run over ANY floor from
 * either generator branch without dragging in decoration.
 */
import { type Grid, type TilePos, at, idx, isWalkable, T_WALL } from "./generator";
import { bfsDistances } from "../engine/flow-field";

/** Coarse region size, in tiles, for the coverage sweep. */
const REGION = 24;
/** A region below this many floor tiles is a sliver and is not held to coverage. */
const REGION_MIN_FLOOR = 120;

export interface FloorMetrics {
  /** Total tiles in the grid, walls included. */
  tiles: number;
  /** Walkable tiles. */
  walkable: number;
  /** walkable ÷ tiles — how much of the rectangle is actually a floor. */
  openShare: number;
  /**
   * Walkable tiles reachable from `start` ÷ walkable. THE hard constraint: a
   * player who can see a region and cannot enter it is the worst thing this
   * generator can ship, and anything below 1 means a pass carved a pocket.
   */
  reachShare: number;
  /** BFS distance start → stairs, in tiles. −1 if the exit is unreachable. */
  pathLen: number;
  /**
   * euclid(start, stairs) ÷ pathLen. 1.0 is a dead-straight shot down a
   * corridor; the lower it is, the more the route has to work for it.
   */
  directness: number;
  /**
   * Direction changes along the traced route ÷ its length. Distinguishes "bent
   * once around a corner" (low directness, low turn rate — one lazy diagonal)
   * from "genuinely snakes", which directness alone cannot.
   */
  turnRate: number;
  /** Walkable tiles with ≤1 walkable neighbour — corridors to nowhere. */
  deadEnds: number;
  /**
   * Walkable tiles with ≥3 walkable neighbours ÷ walkable. The PCG literature's
   * "choice heuristic": how often the player is actually offered a branch.
   */
  choiceShare: number;
  /**
   * Track tiles ÷ walkable, or 0 with no mask. On a track floor the circuit IS
   * the floor's reason to exist, so its share is a design target and not a
   * curiosity — it decayed 0.30 → 0.12 with depth before anyone measured it.
   */
  laneShare: number;
  /** Coarse regions holding real floor area. */
  regions: number;
  /**
   * Coarse regions holding real floor area that the player can reach ÷
   * `regions`. Catches a whole quadrant sealed off behind a bad pass, which a
   * tile-level reachShare of 0.97 would hide.
   */
  regionReachShare: number;
}

/** Which walkable tiles belong to the floor's signature feature (the circuit). */
export interface LaneMask {
  lane: Uint8Array;
}

/**
 * Trace the start→stairs route by walking DOWNHILL through a distance field
 * measured from the start.
 *
 * Downhill from the exit rather than uphill from the spawn: `dist` is BFS from
 * `start`, so every step toward a strictly smaller value is one tile closer to
 * it, and the walk terminates by construction. Returns the path start-first.
 */
export function traceRoute(g: Grid, start: TilePos, stairs: TilePos, dist: Int32Array): TilePos[] {
  if (dist[idx(g, stairs.i, stairs.j)] < 0) return [];
  const path: TilePos[] = [{ i: stairs.i, j: stairs.j }];
  let cur = { i: stairs.i, j: stairs.j };
  // A BFS path is at most `walkable` long; the bound is a runaway guard only.
  for (let guard = 0; guard < g.w * g.h; guard++) {
    if (cur.i === start.i && cur.j === start.j) break;
    const d = dist[idx(g, cur.i, cur.j)];
    let next: TilePos | null = null;
    for (const [di, dj] of [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ] as const) {
      const ni = cur.i + di;
      const nj = cur.j + dj;
      if (!isWalkable(g, ni, nj)) continue;
      const nd = dist[idx(g, ni, nj)];
      if (nd >= 0 && nd === d - 1) {
        next = { i: ni, j: nj };
        break;
      }
    }
    if (!next) break;
    path.push(next);
    cur = next;
  }
  return path.reverse();
}

/**
 * Measure a finished floor.
 *
 * `mask` is optional so this works on both generator branches — the legacy maze
 * has no circuit and simply reports `laneShare` 0.
 *
 * ONE BFS. `bfsDistances` hands back a shared scratch buffer that the next call
 * overwrites, so everything derived from the field is derived here, before any
 * other caller can invalidate it.
 */
export function measureFloor(
  g: Grid,
  start: TilePos,
  stairs: TilePos,
  mask?: LaneMask,
  opts: { routeFrom?: TilePos } = {},
): FloorMetrics {
  const tiles = g.w * g.h;
  let walkable = 0;
  let deadEnds = 0;
  let choices = 0;
  let lane = 0;

  const floorPerRegion = new Map<number, number>();
  const regW = Math.ceil(g.w / REGION);
  const regionOf = (i: number, j: number): number =>
    Math.floor(j / REGION) * regW + Math.floor(i / REGION);

  for (let j = 0; j < g.h; j++) {
    for (let i = 0; i < g.w; i++) {
      if (!isWalkable(g, i, j)) continue;
      walkable++;
      if (mask?.lane[idx(g, i, j)]) lane++;
      let open = 0;
      for (const [di, dj] of [
        [1, 0],
        [-1, 0],
        [0, 1],
        [0, -1],
      ] as const) {
        if (isWalkable(g, i + di, j + dj)) open++;
      }
      if (open <= 1) deadEnds++;
      if (open >= 3) choices++;
      floorPerRegion.set(regionOf(i, j), (floorPerRegion.get(regionOf(i, j)) ?? 0) + 1);
    }
  }

  const dist = bfsDistances(g, start.i, start.j);
  let reached = 0;
  const reachedRegion = new Set<number>();
  for (let j = 0; j < g.h; j++) {
    for (let i = 0; i < g.w; i++) {
      if (!isWalkable(g, i, j)) continue;
      // −1 is `bfsDistances`' unreachable sentinel. Testing `d < BIG` instead
      // counts every unreachable tile as reached and pins the metric at 1.0 —
      // a gate that can only ever pass.
      if (dist[idx(g, i, j)] < 0) continue;
      reached++;
      reachedRegion.add(regionOf(i, j));
    }
  }

  // ── WHERE THE JOURNEY STARTS ────────────────────────────────────────────
  //
  // Reachability is measured from the SPAWN (above — the player is really
  // standing there). The route metrics are measured from `routeFrom`, which on
  // a floor with a launch chute is the chute's MOUTH.
  //
  // This is not the gate being relaxed to accommodate the chute, and the
  // measurement says so: over 240 floors, mean directness from the mouth came
  // out at 0.639 against 0.628 from the base — very slightly HIGHER, and
  // pathLen is strictly shorter, which makes the `minPathSpan` check stricter
  // rather than looser. What it removes is an artefact. The chute is a fixed,
  // deliberately dead-straight ~20-tile lane present on every floor by
  // construction; it adds the same length to euclid and to pathLen, so it drags
  // `euclid/pathLen` toward 1.0 no matter how much the maze beyond it snakes.
  // Counting a non-navigational constant inside a "how hard is this to
  // navigate" ratio measures the wrong span — the same class of error as the
  // `d < BIG` reach bug this file's header records.
  const from = opts.routeFrom ?? start;
  const routeDist = from === start ? dist : bfsDistances(g, from.i, from.j);
  const pathLen = routeDist[idx(g, stairs.i, stairs.j)] ?? -1;
  const route = traceRoute(g, from, stairs, routeDist);
  let turns = 0;
  for (let t = 2; t < route.length; t++) {
    const ai = route[t - 1].i - route[t - 2].i;
    const aj = route[t - 1].j - route[t - 2].j;
    const bi = route[t].i - route[t - 1].i;
    const bj = route[t].j - route[t - 1].j;
    if (ai !== bi || aj !== bj) turns++;
  }
  const euclid = Math.hypot(stairs.i - from.i, stairs.j - from.j);

  const bigRegions = [...floorPerRegion.entries()].filter(([, n]) => n >= REGION_MIN_FLOOR);
  const bigReached = bigRegions.filter(([r]) => reachedRegion.has(r)).length;

  return {
    tiles,
    walkable,
    openShare: walkable / tiles,
    reachShare: walkable ? reached / walkable : 0,
    pathLen,
    directness: pathLen > 0 ? euclid / pathLen : 0,
    turnRate: route.length > 2 ? turns / (route.length - 1) : 0,
    deadEnds,
    choiceShare: walkable ? choices / walkable : 0,
    laneShare: walkable ? lane / walkable : 0,
    regions: bigRegions.length,
    regionReachShare: bigRegions.length ? bigReached / bigRegions.length : 1,
  };
}

/**
 * The constraint band a floor must satisfy. These are the §4 non-negotiables
 * written as numbers; the smoke gate asserts `checkFloor` returns empty.
 *
 * Bands are deliberately WIDE. A gate that pins the current output exactly is a
 * change detector, not a constraint — it goes red on every legitimate tuning
 * pass and gets deleted. These are the walls of the corridor the generator may
 * wander inside, sized from a 200-floor census with headroom.
 */
export interface FloorConstraints {
  minReachShare: number;
  minRegionReachShare: number;
  /** Critical path as a fraction of the grid's Manhattan span. */
  minPathSpan: number;
  maxDirectness: number;
  /** Dead ends allowed per 1000 walkable tiles. */
  maxDeadEndsPer1k: number;
  minOpenShare: number;
  maxOpenShare: number;
  minChoiceShare: number;
}

export const DEFAULT_CONSTRAINTS: FloorConstraints = {
  // Anything under 1 means a carve or an uncarve stranded a pocket. There is no
  // acceptable non-zero rate here, so the band has no slack.
  minReachShare: 1,
  minRegionReachShare: 1,
  // The exit must be a trek, not a doorstep: at least a fifth of the floor's
  // own Manhattan span away along the walkable surface.
  minPathSpan: 0.2,
  // 1.0 is a straight corridor from spawn to stairs. Censused floors run
  // 0.54–0.70, so 0.85 fails only a route that genuinely stopped bending.
  maxDirectness: 0.85,
  // The repair passes hold this near zero (0.3–1.0 per floor, i.e. well under
  // 1 per 1000 tiles); this catches a regression that turns them off, not
  // ordinary variation.
  maxDeadEndsPer1k: 2.5,
  // Below ~0.35 the floor is solid rock with slots cut in it; above ~0.8 there
  // is no maze left and the level is one open blob (a real failure mode — one
  // early track build measured 97% open).
  minOpenShare: 0.35,
  maxOpenShare: 0.8,
  // Some branching must exist or the floor is a single corridor.
  minChoiceShare: 0.2,
};

/** Human-readable constraint violations. Empty array = the floor is legal. */
export function checkFloor(
  m: FloorMetrics,
  g: Grid,
  c: FloorConstraints = DEFAULT_CONSTRAINTS,
): string[] {
  const bad: string[] = [];
  const span = g.w + g.h;
  if (m.reachShare < c.minReachShare)
    bad.push(`unreachable floor: reachShare ${m.reachShare.toFixed(4)} < ${c.minReachShare}`);
  if (m.regionReachShare < c.minRegionReachShare)
    bad.push(
      `sealed region: regionReachShare ${m.regionReachShare.toFixed(3)} < ${c.minRegionReachShare}`,
    );
  if (m.pathLen < 0) bad.push("no route from spawn to stairs");
  else if (m.pathLen < span * c.minPathSpan)
    bad.push(`exit on the doorstep: pathLen ${m.pathLen} < ${(span * c.minPathSpan).toFixed(0)}`);
  if (m.directness > c.maxDirectness)
    bad.push(
      `straight shot to the exit: directness ${m.directness.toFixed(3)} > ${c.maxDirectness}`,
    );
  const dePer1k = m.walkable ? (m.deadEnds * 1000) / m.walkable : 0;
  if (dePer1k > c.maxDeadEndsPer1k)
    bad.push(
      `corridors to nowhere: ${dePer1k.toFixed(2)} dead ends per 1k tiles > ${c.maxDeadEndsPer1k}`,
    );
  if (m.openShare < c.minOpenShare)
    bad.push(`solid rock: openShare ${m.openShare.toFixed(3)} < ${c.minOpenShare}`);
  if (m.openShare > c.maxOpenShare)
    bad.push(`no maze left: openShare ${m.openShare.toFixed(3)} > ${c.maxOpenShare}`);
  if (m.choiceShare < c.minChoiceShare)
    bad.push(`no branches: choiceShare ${m.choiceShare.toFixed(3)} < ${c.minChoiceShare}`);
  return bad;
}

/** One-line census row — for tuning scripts and failure messages. */
export function formatMetrics(m: FloorMetrics): string {
  return [
    `tiles=${m.tiles}`,
    `open=${m.openShare.toFixed(3)}`,
    `reach=${m.reachShare.toFixed(4)}`,
    `path=${m.pathLen}`,
    `direct=${m.directness.toFixed(3)}`,
    `turn=${m.turnRate.toFixed(3)}`,
    `dead=${m.deadEnds}`,
    `choice=${m.choiceShare.toFixed(3)}`,
    `lane=${m.laneShare.toFixed(3)}`,
    `regions=${m.regions}`,
  ].join(" ");
}

/** Tiles that are wall — exported so callers can sanity-check a grid cheaply. */
export function wallCount(g: Grid): number {
  let n = 0;
  for (let k = 0; k < g.t.length; k++) if (at(g, k % g.w, Math.floor(k / g.w)) === T_WALL) n++;
  return n;
}

File: crates/pk-core/src/maze/floor_metrics.rs
//! FLOOR METRICS — how we judge a generated floor, as numbers.
//!
//! Port of `legacy/src/game/pinball-knight/maze/floor-metrics.ts` (433 lines).
//!
//! PORTS: `maze/floor-metrics.ts`

use std::collections::{HashMap, HashSet};

use super::track_launch::TilePos;
use crate::flow_field::bfs_distances;
use crate::grid::{at, idx, is_walkable, Grid, T_WALL};

/// Coarse region size, in tiles, for the coverage sweep.
pub const REGION: usize = 24;

/// A region below this many floor tiles is a sliver and is not held to coverage.
pub const REGION_MIN_FLOOR: usize = 120;

/// Half-width of the neighbourhood a tile must have entirely open to count as "inside a room" (5x5).
pub const CHAMBER_R: i32 = 2;

/// Which walkable tiles belong to the floor's signature feature (the circuit).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LaneMask {
    pub lane: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FloorMetrics {
    /// Total tiles in the grid, walls included.
    pub tiles: usize,
    /// Walkable tiles.
    pub walkable: usize,
    /// walkable ÷ tiles — how much of the rectangle is actually a floor.
    pub open_share: f64,
    /// Walkable tiles reachable from start ÷ walkable.
    pub reach_share: f64,
    /// BFS distance start → stairs, in tiles. -1 if exit is unreachable.
    pub path_len: i32,
    /// euclid(start, stairs) ÷ path_len.
    pub directness: f64,
    /// Direction changes along the traced route ÷ its length.
    pub turn_rate: f64,
    /// Walkable tiles with <= 1 walkable neighbour — corridors to nowhere.
    pub dead_ends: usize,
    /// Walkable tiles with >= 3 walkable neighbours ÷ walkable.
    pub choice_share: f64,
    /// Track tiles ÷ walkable, or 0 with no mask.
    pub lane_share: f64,
    /// The floor's biggest actual room ÷ walkable (connected 5x5 open block).
    pub chamber_share: f64,
    /// Coarse regions holding real floor area (>= REGION_MIN_FLOOR).
    pub regions: usize,
    /// Coarse regions holding real floor area that player can reach ÷ regions.
    pub region_reach_share: f64,
}

/// The largest connected blob of tiles that are deep inside open space, in tiles.
pub fn largest_chamber(g: &Grid) -> usize {
    let n_tiles = (g.w * g.h) as usize;
    let mut deep = vec![0_u8; n_tiles];

    for j in CHAMBER_R..(g.h - CHAMBER_R) {
        for i in CHAMBER_R..(g.w - CHAMBER_R) {
            let mut ok = 1_u8;
            'outer: for dj in -CHAMBER_R..=CHAMBER_R {
                for di in -CHAMBER_R..=CHAMBER_R {
                    if !is_walkable(g, i + di, j + dj) {
                        ok = 0;
                        break 'outer;
                    }
                }
            }
            deep[idx(g, i, j)] = ok;
        }
    }

    let mut seen = vec![0_u8; n_tiles];
    let mut stack = Vec::new();
    let mut best = 0;

    for k in 0..n_tiles {
        if deep[k] == 0 || seen[k] == 1 {
            continue;
        }
        let mut count = 0;
        stack.clear();
        stack.push(k);
        seen[k] = 1;

        while let Some(c) = stack.pop() {
            count += 1;
            let ci = (c % (g.w as usize)) as i32;
            let cj = (c / (g.w as usize)) as i32;

            for (di, dj) in &[(1, 0), (-1, 0), (0, 1), (0, -1)] {
                let ni = ci + di;
                let nj = cj + dj;
                if ni < 0 || nj < 0 || ni >= g.w || nj >= g.h {
                    continue;
                }
                let nk = idx(g, ni, nj);
                if seen[nk] == 1 || deep[nk] == 0 {
                    continue;
                }
                seen[nk] = 1;
                stack.push(nk);
            }
        }
        if count > best {
            best = count;
        }
    }

    best
}

/// Trace the start -> stairs route by walking downhill through a distance field.
pub fn trace_route(g: &Grid, start: TilePos, stairs: TilePos, dist: &[i32]) -> Vec<TilePos> {
    if dist[idx(g, stairs.i, stairs.j)] < 0 {
        return Vec::new();
    }
    let mut path = vec![TilePos {
        i: stairs.i,
        j: stairs.j,
    }];
    let mut cur = TilePos {
        i: stairs.i,
        j: stairs.j,
    };

    let max_guard = (g.w * g.h) as usize;
    for _ in 0..max_guard {
        if cur.i == start.i && cur.j == start.j {
            break;
        }
        let d = dist[idx(g, cur.i, cur.j)];
        let mut next = None;

        for (di, dj) in &[(1, 0), (-1, 0), (0, 1), (0, -1)] {
            let ni = cur.i + di;
            let nj = cur.j + dj;
            if ni < 0 || nj < 0 || ni >= g.w || nj >= g.h {
                continue;
            }
            if !is_walkable(g, ni, nj) {
                continue;
            }
            let nd = dist[idx(g, ni, nj)];
            if nd >= 0 && nd == d - 1 {
                next = Some(TilePos { i: ni, j: nj });
                break;
            }
        }
        match next {
            Some(nxt) => {
                path.push(nxt);
                cur = nxt;
            }
            None => break,
        }
    }

    path.reverse();
    path
}

/// Measure a finished floor.
pub fn measure_floor(
    g: &Grid,
    start: TilePos,
    stairs: TilePos,
    mask: Option<&LaneMask>,
    route_from: Option<TilePos>,
) -> FloorMetrics {
    let tiles = (g.w * g.h) as usize;
    let mut walkable = 0;
    let mut dead_ends = 0;
    let mut choices = 0;
    let mut lane = 0;

    let mut floor_per_region: HashMap<usize, usize> = HashMap::new();
    let reg_w = (g.w as f64 / REGION as f64).ceil() as usize;
    let region_of = |i: i32, j: i32| -> usize {
        let rj = (j as usize) / REGION;
        let ri = (i as usize) / REGION;
        rj * reg_w + ri
    };

    for j in 0..g.h {
        for i in 0..g.w {
            if !is_walkable(g, i, j) {
                continue;
            }
            walkable += 1;
            let k = idx(g, i, j);
            if let Some(m) = mask {
                if k < m.lane.len() && m.lane[k] == 1 {
                    lane += 1;
                }
            }
            let mut open = 0;
            for (di, dj) in &[(1, 0), (-1, 0), (0, 1), (0, -1)] {
                let ni = i + di;
                let nj = j + dj;
                if ni >= 0 && nj >= 0 && ni < g.w && nj < g.h && is_walkable(g, ni, nj) {
                    open += 1;
                }
            }
            if open <= 1 {
                dead_ends += 1;
            }
            if open >= 3 {
                choices += 1;
            }
            *floor_per_region.entry(region_of(i, j)).or_insert(0) += 1;
        }
    }

    let dist = bfs_distances(g, start.i, start.j);
    let mut reached = 0;
    let mut reached_region: HashSet<usize> = HashSet::new();

    for j in 0..g.h {
        for i in 0..g.w {
            if !is_walkable(g, i, j) {
                continue;
            }
            if dist[idx(g, i, j)] < 0 {
                continue;
            }
            reached += 1;
            reached_region.insert(region_of(i, j));
        }
    }

    let from = route_from.unwrap_or(start);
    let route_dist_owned;
    let route_dist = if from == start {
        &dist
    } else {
        route_dist_owned = bfs_distances(g, from.i, from.j);
        &route_dist_owned
    };

    let path_len = route_dist[idx(g, stairs.i, stairs.j)];
    let route = trace_route(g, from, stairs, route_dist);
    let mut turns = 0;
    if route.len() > 2 {
        for t in 2..route.len() {
            let ai = route[t - 1].i - route[t - 2].i;
            let aj = route[t - 1].j - route[t - 2].j;
            let bi = route[t].i - route[t - 1].i;
            let bj = route[t].j - route[t - 1].j;
            if ai != bi || aj != bj {
                turns += 1;
            }
        }
    }
    let di = (stairs.i - from.i) as f64;
    let dj = (stairs.j - from.j) as f64;
    let euclid = (di * di + dj * dj).sqrt();

    let big_regions: Vec<(&usize, &usize)> = floor_per_region
        .iter()
        .filter(|(_, &n)| n >= REGION_MIN_FLOOR)
        .collect();
    let big_reached = big_regions
        .iter()
        .filter(|(&r, _)| reached_region.contains(&r))
        .count();

    FloorMetrics {
        tiles,
        walkable,
        open_share: if tiles > 0 {
            walkable as f64 / tiles as f64
        } else {
            0.0
        },
        reach_share: if walkable > 0 {
            reached as f64 / walkable as f64
        } else {
            0.0
        },
        path_len,
        directness: if path_len > 0 {
            euclid / path_len as f64
        } else {
            0.0
        },
        turn_rate: if route.len() > 2 {
            turns as f64 / (route.len() - 1) as f64
        } else {
            0.0
        },
        dead_ends,
        choice_share: if walkable > 0 {
            choices as f64 / walkable as f64
        } else {
            0.0
        },
        lane_share: if walkable > 0 {
            lane as f64 / walkable as f64
        } else {
            0.0
        },
        chamber_share: if walkable > 0 {
            largest_chamber(g) as f64 / walkable as f64
        } else {
            0.0
        },
        regions: big_regions.len(),
        region_reach_share: if !big_regions.is_empty() {
            big_reached as f64 / big_regions.len() as f64
        } else {
            1.0
        },
    }
}

/// The constraint band a floor must satisfy.
#[derive(Debug, Clone, PartialEq)]
pub struct FloorConstraints {
    pub min_reach_share: f64,
    pub min_region_reach_share: f64,
    /// Critical path as a fraction of the grid's Manhattan span.
    pub min_path_span: f64,
    pub max_directness: f64,
    /// Dead ends allowed per 1000 walkable tiles.
    pub max_dead_ends_per_1k: f64,
    pub min_open_share: f64,
    pub max_open_share: f64,
    pub min_choice_share: f64,
}

pub const DEFAULT_CONSTRAINTS: FloorConstraints = FloorConstraints {
    min_reach_share: 1.0,
    min_region_reach_share: 1.0,
    min_path_span: 0.2,
    max_directness: 0.85,
    max_dead_ends_per_1k: 2.5,
    min_open_share: 0.35,
    max_open_share: 0.8,
    min_choice_share: 0.2,
};

/// Human-readable constraint violations. Empty Vec = the floor is legal.
pub fn check_floor(m: &FloorMetrics, g: &Grid, c: &FloorConstraints) -> Vec<String> {
    let mut bad = Vec::new();
    let span = (g.w + g.h) as f64;

    if m.reach_share < c.min_reach_share {
        bad.push(format!(
            "unreachable floor: reachShare {:.4} < {}",
            m.reach_share, c.min_reach_share
        ));
    }
    if m.region_reach_share < c.min_region_reach_share {
        bad.push(format!(
            "sealed region: regionReachShare {:.3} < {}",
            m.region_reach_share, c.min_region_reach_share
        ));
    }
    if m.path_len < 0 {
        bad.push("no route from spawn to stairs".into());
    } else if (m.path_len as f64) < span * c.min_path_span {
        bad.push(format!(
            "exit on the doorstep: pathLen {} < {:.0}",
            m.path_len,
            span * c.min_path_span
        ));
    }
    if m.directness > c.max_directness {
        bad.push(format!(
            "straight shot to the exit: directness {:.3} > {}",
            m.directness, c.max_directness
        ));
    }
    let de_per_1k = if m.walkable > 0 {
        (m.dead_ends as f64 * 1000.0) / m.walkable as f64
    } else {
        0.0
    };
    if de_per_1k > c.max_dead_ends_per_1k {
        bad.push(format!(
            "corridors to nowhere: {:.2} dead ends per 1k tiles > {}",
            de_per_1k, c.max_dead_ends_per_1k
        ));
    }
    if m.open_share < c.min_open_share {
        bad.push(format!(
            "solid rock: openShare {:.3} < {}",
            m.open_share, c.min_open_share
        ));
    }
    if m.open_share > c.max_open_share {
        bad.push(format!(
            "no maze left: openShare {:.3} > {}",
            m.open_share, c.max_open_share
        ));
    }
    if m.choice_share < c.min_choice_share {
        bad.push(format!(
            "no branches: choiceShare {:.3} < {}",
            m.choice_share, c.min_choice_share
        ));
    }

    bad
}

/// One-line census row — for tuning scripts and failure messages.
pub fn format_metrics(m: &FloorMetrics) -> String {
    format!(
        "tiles={} open={:.3} reach={:.4} path={} direct={:.3} turn={:.3} dead={} choice={:.3} lane={:.3} chamber={:.3} regions={}",
        m.tiles,
        m.open_share,
        m.reach_share,
        m.path_len,
        m.directness,
        m.turn_rate,
        m.dead_ends,
        m.choice_share,
        m.lane_share,
        m.chamber_share,
        m.regions
    )
}

/// Tiles the player can stand on — the number every density budget should ride.
pub fn walkable_count(g: &Grid) -> usize {
    let mut n = 0;
    for j in 0..g.h {
        for i in 0..g.w {
            if is_walkable(g, i, j) {
                n += 1;
            }
        }
    }
    n
}

/// Tiles that are wall — exported so callers can sanity-check a grid cheaply.
pub fn wall_count(g: &Grid) -> usize {
    let mut n = 0;
    for k in 0..g.t.len() {
        let i = (k % g.w as usize) as i32;
        let j = (k / g.w as usize) as i32;
        if at(g, i, j) == T_WALL {
            n += 1;
        }
    }
    n
}
